components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt

```diff
@@ -145,6 +170,38 @@ open class GleanInternalAPI internal constructor () {
         return uploadEnabled
     }
 
+    /**
+     * Clear any pending metrics when telemetry is disabled.
+     */
+    @Suppress("EXPERIMENTAL_API_USAGE")
+    private fun clearMetrics() = Dispatchers.API.launch {
+        // There is only one metric that we want to survive after clearing all metrics:
+        // firstRunDate. Here, we store its value so we can restore it after clearing
+        // the metrics.
+        val firstRunDateMetric = GleanInternalMetrics.firstRunDate
+        val existingFirstRunDate = DatetimesStorageEngine.getSnapshot(
+            firstRunDateMetric.sendInPings[0], false)?.get(firstRunDateMetric.identifier)?.let {
+            parseISOTimeString(it)
+        }
+
+        pingStorageEngine.clearPendingPings()
```

Comment:
It just occurred to me that we also need to cancel the workmanager upload job, probably before clearing the ping storage. What do you think, @mdboom ?

Reply:
That's a good point.  

(1) There's probably a slight chance of a race condition that pings will get sent right before we do this (but sending and deleting pings is at least synchronized so hopefully that doesn't cause an I/O error reading a file that was just deleted etc.).

(2) The WorkManager will keep waking up, finding no pings, and doing nothing.  Which is suboptimal, but doesn't violate the privacy expectations, at least.

Anyway, it's worth figuring out how to turn on/off the workmanager as part of the `uploadEnabled` flag.

Reply:
> 
> 
> That's a good point.
> 
> (1) There's probably a slight chance of a race condition that pings will get sent right before we do this (but sending and deleting pings is at least synchronized so hopefully that doesn't cause an I/O error reading a file that was just deleted etc.).

Yes. I think we could still live with the I/O error, if it happens, in this case.

> (2) The WorkManager will keep waking up, finding no pings, and doing nothing. Which is suboptimal, but doesn't violate the privacy expectations, at least.

Would it? We bail out attempting to send pings if upload is disabled. [See here](https://github.com/mozilla-mobile/android-components/blob/786e1a30d32e3b89c54f057523cd8b1616c8c50d/components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt#L358).

> Anyway, it's worth figuring out how to turn on/off the workmanager as part of the `uploadEnabled` flag.

I agree. I think this could be some method on the PingWorker that [cancels the worker](https://developer.android.com/topic/libraries/architecture/workmanager/how-to/cancel-stop-work#stopping).

Reply:
@mdboom I'm also happy to take this to a follow-up PR, doesn't need to block this.

Reply:
It was easy enough to add, but I'm a little stumped by how to test it.

Reply:
What about, for now, simply asserting `Assert.assertFalse("A scheduled PingUploadWorker must exist", isWorkScheduled(PingUploadWorker.PING_WORKER_TAG))`

Reply:
That makes sense.

Reply:
This is turning into a confusing mess of async WorkManager stuff and NPEs, so I'm going to pass on this for now and make a new bug to deal with this down the road.